File: 0120-triangle/0120-triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        #Tabulation  TC-O(N*N),SC-O(N*N)
        
        n=len(triangle)
        dp=[[0]*n for _ in range(n)]
        for j in range(n):
            dp[n-1][j]=triangle[n-1][j]

        for i in range(n-2,-1,-1):
            for j in range(i,-1,-1):
                down=triangle[i][j]+dp[i+1][j]
                dia=triangle[i][j]+dp[i+1][j+1]
                dp[i][j]=min(down,dia)


        return dp[0][0]        

File: 0120-triangle/0120-triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int n=triangle.size();
        int dp[][]=new int[n][n];
        for(int i=0;i<n;i++){
            dp[n-1][i]=triangle.get(n-1).get(i);
        }
        for(int i=n-2;i>=0;i--){
            for(int j=0;j<=i;j++){
                int right=dp[i+1][j+1];
                int down=dp[i+1][j];
                dp[i][j]=Math.min(right,down)+triangle.get(i).get(j);
            }
        }
        return dp[0][0];
        
        



    }}